Refactor image builds to use buildx for multi arch image building

RobReus opened the pull request:
Fixes #2466 , Fixes #2486 , Closes #2566 

This change introduces [buildx](https://docs.docker.com/buildx/working-with-buildx/) to build multi arch container images.

Only the container image building is affected by this change, all other functionality (including locally building) remains the same.

Would love to get some feedback on this. We would probably improve it further to reduce build times. Right now the GitHub workflow takes about the same amount of time as before, even though images are now built in parallel. I noticed that cross compiling within the new way of doing it went from 90 seconds to around 300 seconds. I feel like we can win plenty of time by making the cross compile of the velero binary more efficient.

Succesful workflow on my fork: https://github.com/RobReus/velero/actions/runs/180022436 (most recent ones failed due to my credentials now having access to the velero registry).

nrb commented on Dockerfile:
```diff
@@ -0,0 +1,67 @@
+# Copyright 2017, 2019 the Velero contributors.
```
Copyright year should be 2020

RobReus replied:
Completely forgot about that, thanks for the feedback. Fixed now.

nrb commented:
Thanks for this @RobReus! I'll look at it in more detail later today. For now, can you double check to make sure that all changed files have their copyright year set to 2020?

RobReus commented:
@nrb I have updated the copyright of the files I touched with this PR.

ashish-amarnath commented on Makefile:
```diff
@@ -43,16 +46,25 @@ VERSION ?= main
 CLI_PLATFORMS ?= linux-amd64 linux-arm linux-arm64 darwin-amd64 windows-amd64 linux-ppc64le
-CONTAINER_PLATFORMS ?= linux-amd64 linux-ppc64le linux-arm linux-arm64
-MANIFEST_PLATFORMS ?= amd64 ppc64le arm arm64
+BUILDX_PLATFORMS ?= linux/amd64,linux/arm64,linux/arm/v7,linux/ppc64le
```
does `buildx` also support building windows containers should we need to build one in the future?

RobReus replied:
I don't have any experience working with Windows container images, but is that really needed? Windows can run `linux/amd64` (in the event the Windows arch is x86_64) without any issues (as far as I know), so there is no need for a Windows based image.

RobReus commented on Makefile:
```diff
@@ -175,47 +144,29 @@ shell: build-dirs build-env
+	--build-arg=GIT_TREE_STATE=$(GIT_TREE_STATE) \
+	-f Dockerfile .
+
+container:
```
The default buildx builder uses the `docker` driver, which does indeed not support auto-push. The Workflow environment uses the `docker-container` driver. You can configure your local environment to use `docker-container` as well by doing this:
`docker buildx create --name builder --driver docker-container --use` and then:
`docker buildx inspect --bootstrap` to verify the builder is configured properly.
See also https://medium.com/@artur.klauser/building-multi-architecture-docker-images-with-buildx-27d80f7e2408 for more information. I am using the docker image based installation, and not the host based installation.

I am running on Linux, so I can't say if there are any special steps to take on Darwin.

RobReus replied:
Alternatively, you can just remove the `--output=type=registry` from the `Makefile` on line 158. All the layers will still be built, but not pushed. However, due to the way buildx works, you will also not have an image as the docker daemon currently has no support for loading a multi-arch image, so `--load` doesn't work. What you might be able to do is output to "local" so the image is written as a file. See: https://github.com/docker/buildx#-o---outputpath-typetypekeyvalue

ashish-amarnath replied:
OK so the instructions seem to be the same for darwin as well.
```
docker buildx create ...
```
We'll want these instructions documented. https://velero.io/docs/v1.4/development/ would be a good place for it.

After that is done, I am seeing this error
```
 => ERROR [linux/ppc64le stage-2 2/3] RUN apt-get update && apt-get install -y ca-certificates && rm -rf /var/lib/apt/lists/*                         1.6s
 => CANCELED [linux/arm/v7 stage-2 2/3] RUN apt-get update && apt-get install -y ca-certificates && rm -rf /var/lib/apt/lists/*                       0.6s
 => CANCELED [linux/amd64 stage-2 2/3] RUN apt-get update && apt-get install -y ca-certificates && rm -rf /var/lib/apt/lists/*                        0.4s
------
 > [linux/ppc64le stage-2 2/3] RUN apt-get update && apt-get install -y ca-certificates && rm -rf /var/lib/apt/lists/*:
#20 1.242 qemu: uncaught target signal 4 (Illegal instruction) - core dumped
#20 1.251 Illegal instruction
------
failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c apt-get update && apt-get install -y ca-certificates && rm -rf /var/lib/apt/lists/*]: buildkit-runc did not terminate successfully
```
Any ideas on what this might me?

RobReus replied:
@ashish-amarnath sounds like OSX does not play nicely with the ppc64le image. What is the output of `docker buildx inspect --bootstrap`?

ashish-amarnath replied:
Debugged this offline w/ @RobReus.
Building for `linux/ppc64le` doesn't work on OSX. We might have to exclude `linux/ppc64le` from building on OSX but build that in the CI.

RobReus commented on hack/download-restic.sh:
```diff
@@ -0,0 +1,54 @@
+    exit 1
+fi
+
+if [[ "${BIN}" != "velero" ]]; then
```
In the old setup, the `restic` binary was only downloaded into the `Docker-velero[-$ARCH]` images. This is why I am checking if the current binary being built is velero, if not, it will not download restic. I don't mind changing stuff about this, as I was just mimicking what was included before to ensure 100% compatibility.

RobReus replied:
Example of before: https://github.com/vmware-tanzu/velero/blob/main/Dockerfile-velero#L21

The reason I am doing it like this now is so we can use a single `Dockerfile` for all the architectures and binaries. Without this check, the restic binary would also end up in the `velero-restic-restore-helper` image, which it wasn't in before. It probably would not break anything, but this was just me making sure the end-result is still 100% the same :)

RobReus commented on hack/download-restic.sh:
```diff
@@ -0,0 +1,54 @@
+
+if [[ "${BIN}" != "velero" ]]; then
+    echo "${BIN} does not need the restic binary"
+    exit 0
```
No, if the `$BIN` is not `velero`, the script should exit gracefully so the `Dockerfile` continues to be built without `restic` (which is the case for the `velero-restic-restore-helper` image).

ashish-amarnath commented on Makefile:
```diff
@@ -175,47 +144,29 @@ shell: build-dirs build-env
+
+container:
+	@docker buildx build --pull \
+	--output=type=registry \
```
Is there a way to push images outside of the auto-push?
This is to avoid accidental pushes to the `velero` registry and also most of our dev work happens locally without an image registry.

RobReus replied:
There is not right now, it will be possible once https://github.com/moby/moby/pull/38738 has been merged and released.

As a work around in the mean time, I have added a check to the container build recipe to prevent accidental unintended pushes and added a local build option for the container images, which will only build the velero and velero-restic-restore-helper images for the arch you are running it from. This will allow a developer to locally build a docker image so it can be tested.

Let me know if this solution works for you.

ashish-amarnath replied:
That is reasonable to me!

ashish-amarnath commented on Makefile:
```diff
@@ -175,46 +159,37 @@ shell: build-dirs build-env
-	docker tag $(IMAGE):$(VERSION) $(IMAGE):latest
-	docker push $(IMAGE):latest
+container-builder-env:
+ifneq ($(BUILDX_ENABLED), true)
```
suggest adding instructions on how to enable this.
Printing a link to the docs should be suffiecient.

RobReus replied:
Agreed, I will be adding some documentation regarding buildx, and link to it here.

ashish-amarnath commented on Makefile:
```diff
@@ -43,16 +46,34 @@ VERSION ?= main
-CONTAINER_PLATFORMS ?= linux-amd64 linux-ppc64le linux-arm linux-arm64
-MANIFEST_PLATFORMS ?= amd64 ppc64le arm arm64
+BUILDX_PLATFORMS ?= linux/amd64,linux/arm64,linux/arm/v7,linux/ppc64le
+BUILDX_OUTPUT_TYPE ?= registry
```
Can we make this default to `docker` and use `registry` only from [`hack/docker-push.sh`](https://github.com/vmware-tanzu/velero/blob/main/hack/docker-push.sh)?
Same with the `BUILDX_PLATFORMS`.

That would avoid creating the `local-all-containers` make target. WDYT?

RobReus replied:
That sounds reasonable to me, I will refactor.

RobReus commented:
@ashish-amarnath I have updated the contribution docs on building the container images. I have also implemented the feedback given.

Let me know if there's anything else you want to discuss or see changed.

nrb commented on Dockerfile:
```diff
@@ -0,0 +1,67 @@
+
+ENV CGO_ENABLED=0 \
+    GO111MODULE=on \
+    GOPROXY=https://proxy.golang.org \
```
Is there any way to overide this variable in at build time? I ask because there are times that Velero will be built within a private environment, without access to the internet, and an internal goproxy server should be used.

Perhaps we introduce another `ARG`?

https://vsupalov.com/docker-build-pass-environment-variables/ also lets us use the hosts's GOPROXY env var value at build time, which I think would be useful as a shortcut, but I think making it explicit is easier to follow later on.

RobReus replied:
@nrb fair point. I just pushed a commit which makes `GOPROXY` for both the regular build-image as the multi-arch build image configurable with the default set to `https://proxy.golang.org`.
